app/assets/javascripts/preferences/panes/two-factor-auth/TwoFactorAuthView.tsx

```diff
@@ -0,0 +1,135 @@
+import { FunctionalComponent, FunctionComponent } from 'preact';
+import {
+  Title,
+  Text,
+  PreferencesGroup,
+  PreferencesSegment,
+} from '../../components';
+import { Switch } from '../../../components/Switch';
+import { observer } from 'mobx-react-lite';
+import { DecoratedInput } from '../../../components/DecoratedInput';
+import { IconButton } from '../../../components/IconButton';
+import { ScanQRCode } from './ScanQRCode';
+import { SaveSecretKey } from './SaveSecretKey';
+import { Verification } from './Verification';
+import { TwoFactorActivation, TwoFactorAuth } from './model';
+import { downloadSecretKey } from './download-secret-key';
+import { CircleProgress } from '@/components/CircleProgress';
+import { useState } from 'preact/hooks';
+import { useEffect } from 'react';
+
+export const TwoFactorAuthView: FunctionComponent<{
+  auth: TwoFactorAuth;
+}> = observer(({ auth }) => (
+  <PreferencesGroup>
+    <PreferencesSegment>
+      <div className="flex flex-row items-center">
+        <div className="flex-grow flex flex-col">
+          <Title>Two-factor authentication</Title>
+          <Text>
+            An extra layer of security when logging in to your account.
+          </Text>
+        </div>
+        <Switch
+          checked={auth.enabled !== false}
+          onChange={() => auth.toggle2FA()}
+        />
+      </div>
+    </PreferencesSegment>
+    <PreferencesSegment>
+      {auth.enabled !== false && (
+        <TwoFactorEnabledView
+          secretKey={auth.enabled.secretKey}
+          authCode={auth.enabled.authCode}
+        />
+      )}
+
+      {auth.activation !== false && (
+        <TwoFactorActivationView activation={auth.activation} />
+      )}
+
+      {auth.enabled === false && <TwoFactorDisabledView />}
+    </PreferencesSegment>
+  </PreferencesGroup>
+));
+
+const ProgressTime: FunctionalComponent<{ time: number }> = ({ time }) => {
+  const [percent, setPercent] = useState(0);
+  const interval = time / 100;
+  useEffect(() => {
+    const tick = setInterval(() => {
+      if (percent === 100) {
+        setPercent(0);
+      } else {
+        setPercent(percent + 1);
+      }
+    }, interval);
+    return () => {
+      clearInterval(tick);
+    };
+  });
+  return <CircleProgress percent={percent} />;
+};
+
+const TwoFactorEnabledView: FunctionComponent<{
+  secretKey: string;
+  authCode: string;
+}> = ({ secretKey, authCode }) => {
+  const download = (
+    <IconButton
+      icon="download"
+      onClick={() => {
+        downloadSecretKey(secretKey);
+      }}
+    />
+  );
+  const copy = (
+    <IconButton
+      icon="copy"
+      onClick={() => {
+        navigator?.clipboard?.writeText(secretKey);
+      }}
+    />
+  );
```

Comment:
We could create a separate higher order component that returns receives icon and click handler and returns corresponding `<IconButton />` component.

Reply:
But, isn't that what IconButton already does?

Reply:
I was referring to code duplication related to `IconButton`. For example, we have defined `copy` variable in 3 different places and only `onClick` handler is different there, so we could add abstraction over it. However, after looking into `IconButton` and its children, it seems introducing a new component will really not be so useful here.